### PR TITLE
Adds Cupertino and Material Widget Playgrounds

### PR DESCRIPTION
**Pull request by btobin**
cc/ @mit-mit @xster 

- Updates to Gallery home screen icons and categories
- Adds Material Playground and Cupertino Playground icons, and respective tabbed pages/widgets

### CODE REVIEW

File: examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+  String code;
+}
+
+class IconChoice {
+  IconChoice({@required this.icon, @required this.code});
+  IconData icon;
+  String code;
+}
+
+final List<BorderChoice> borderChoices = <BorderChoice>[
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart]: `const List<BorderChoice> borderChoices = ...` 

Likewise for colorChocies, iconChocies

File: examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    iconChoices.map((IconChoice i) => i.icon).toList();
+
+String codeSnippetForColor(Color color) {
+  return colorChoices
+      .where((ColorChoice c) => c.color == color)
+      .toList()[0]
+      .code;
+}
+
+String codeSnippetForBorder(String type) {
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart]:
```
String codeSnippetForBorder(String type) {
  return borderChoices.firstWhere((BorderChoice b) => b.type == type).code;
}
```

Likewise for `codeSnippetForColor`, `codeSnippetForIcon`

File: examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+      .where((BorderChoice b) => b.type == type)
+      .toList()[0]
+      .code;
+}
+
+String codeSnippetForIcon(IconData icon) {
+  return iconChoices.where((IconChoice b) => b.icon == icon).toList()[0].code;
+}
+
+Widget sliderPicker({
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart]: Factor widgets into more widgets.

```
class SliderPicker extends StatelessWidget {
  const SliderPicker({
    Key key,
    @required this.label,
    this.value = 0.0,
    this.minValue = 0.0,
    this.maxValue = 1.0,
    this.divisions,
    this.onValueChanged,
  }) : assert(label != null), super(key: key);

  final String label;
  final double value;
  final double minValue;
  final double maxValue;
  final int divisions;
  final ValueChanged<double> onValueChanged;

  @override
  Widget builder(BuildContext context) {
    return PropertyColumn(
      label: label,
      widget: Padding(
        padding: const EdgeInsets.only(bottom: 5.0),
        child: Slider(
          value: value,
          min: minValue,
          max: maxValue,
          divisions: divisions,
          onChanged: onValueChanged,
        ),
      ),
    );
  }
}
```

File: examples/flutter_gallery/lib/demo/playground/configuration/property_column.dart [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      mainAxisSize: MainAxisSize.min,
+      mainAxisAlignment: MainAxisAlignment.start,
+      crossAxisAlignment: CrossAxisAlignment.stretch,
+      children: <Widget>[
+        Padding(
+          padding: const EdgeInsets.only(left: 15.0)
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/property_column.dart]: This appears to be `const EdgeInsets.only(left: 15.0, top: 15.0, bottom: 20.0)`

File: examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+  BorderSide borderSide = const BorderSide(
+    color: Colors.grey,
+    width: 2.0,
+  );
+
+  borderSide = side ? borderSide : BorderSide.none;
+
+  switch (type) {
+    case 'square':
+      shape = RoundedRectangleBorder(
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart]: just `return RoundedRectangleBorder(borderRadius: BorderRadius.zero, side: borderSide);`

Likewise for the rest of switch clauses

**Comment by lukef** [comment]
Confirmed. Good to include my code.

File: examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+          side: borderSide.copyWith(width: 1.0));
+      break;
+    case 'stadium':
+      shape = StadiumBorder(side: borderSide);
+      break;
+  }
+  return shape;
+}
+
+Widget shapeButton({
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart]:
```
class ShapeButton extends StatelessWidget {
  const ShapeButton({
    Key key,
    @required this.shape,
    this.size = const Size(_pickerRowHeight, _pickerRowHeight),
    this.isSelected = false,
    this.onTapped,
  }) : assert(shape != null), super(key: key);

  final String shape;
  final Size size;
  final bool isSelected;
  final VoidCallback onTapped;

  @override
  Widget build(BuildContext context) {
    return ButtonTheme(
      minWidth: size.width,
      height: size.height,
      child: RaisedButton(
        shape: borderShapeFromString(shape),
        color: isSelected ? Colors.blue : Colors.white,
        elevation: isSelected ? _pickerSelectedElevation : 0.0,
        onPressed: onTapped,
      ),
    );
  }
}
```

File: examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+  );
+}
+
+Widget shapePicker({
+  double pickerHeight = _pickerRowHeight,
+  String selectedValue,
+  List<String> shapeNames,
+  IndexedValueCallback<String> onItemTapped,
+}) {
+  final List<String> shapeNameOptions = shapeNames ??
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart]: Looks like shapeNames should be:
`enum ShapeNames { square, beveled, rounded, stadium }`

And shapeNameOptions would be:
` final List<String> shapeNameOptions = shapeNames ?? ShapeNames.values`;

File: examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+    child: RaisedButton(
+      shape: borderShapeFromString(shape),
+      color: isSelected ? Colors.blue : Colors.white,
+      elevation: isSelected ? _pickerSelectedElevation : 0.0,
+      onPressed: onTapped,
+    ),
+  );
+}
+
+Widget shapePicker({
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart]: `class ShapePicker extends StatelessWidget` etc

File: examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+        mainAxisSize: MainAxisSize.max,
+        mainAxisAlignment: MainAxisAlignment.center,
+        crossAxisAlignment: CrossAxisAlignment.center,
+        children: buttonChildren,
+      ),
+    ),
+  );
+}
+
+Widget colorButton({
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart]: `class ColorButton extends StatelessWidget` etc

File: examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+  bool isSelected = false,
+  VoidCallback onTapped,
+}) {
+  return ButtonTheme(
+    minWidth: size.width,
+    height: size.height,
+    child: RaisedButton(
+      shape: StadiumBorder(
+        side: BorderSide(
+          color:
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart]: no line break here

File: examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+      splashColor: color == Colors.white
+          ? Colors.grey[400].withOpacity(0.3)
+          : Colors.white.withOpacity(0.3),
+      elevation: isSelected ? _pickerSelectedElevation : 0.0,
+      onPressed: onTapped,
+    ),
+  );
+}
+
+Widget colorPicker({
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart]: class ColorPicker extends StatelessWidget etc

File: examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+        mainAxisSize: MainAxisSize.max,
+        mainAxisAlignment: MainAxisAlignment.center,
+        crossAxisAlignment: CrossAxisAlignment.center,
+        children: buttonChildren,
+      ),
+    ),
+  );
+}
+
+Widget buttonListContainer(List<Widget> children,
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart]: class ButtonListContainer extends StatelessWidget etc

File: examples/flutter_gallery/lib/demo/playground/home.dart [hunk 13]
```diff
@@ -0,0 +1,49 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+import 'playground_demo.dart';
+
+class PlaygroundScaffold extends StatelessWidget {
+  const PlaygroundScaffold({
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/home.dart]: Although it's not necessary here, it's still best to include a `Key, key` parameter and pass it along with `: super(key: key)` for the sake of copy and pasters.

File: examples/flutter_gallery/lib/demo/playground/home.dart [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+                Navigator.maybePop(context);
+              }),
+          bottom: TabBar(
+            isScrollable: true,
+            tabs: demos.map<Widget>((PlaygroundDemo demo) {
+              return Tab(text: demo.tabName());
+            }).toList(),
+          ),
+        ),
+        body: TabBarView(
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/home.dart]: `body: TabBarView(children: demos)`

File: examples/flutter_gallery/lib/demo/playground/home.dart [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return DefaultTabController(
+      length: demos.length,
+      child: Scaffold(
+        key: _scaffoldKey,
+        appBar: AppBar(
+          title: Text(title),
+          leading: IconButton(
+              icon: const BackButtonIcon(),
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/home.dart]: two space indent

File: examples/flutter_gallery/lib/demo/playground/home.dart [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+class PlaygroundScaffold extends StatelessWidget {
+  const PlaygroundScaffold({
+    this.title,
+    this.demos,
+  });
+
+  final String title;
+  final List<PlaygroundDemo> demos;
+
+  static final GlobalKey<ScaffoldState> _scaffoldKey =
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/home.dart]: Is this key really needed? If it is, please provide a comment (and remove the line break).

File: examples/flutter_gallery/lib/demo/playground/cupertino/segment_control.dart [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/material.dart';
+
+import '../configuration/material_helpers.dart';
+import '../playground_demo.dart';
+
+const String _demoWidgetName = 'CupertinoSegmentControl';
+
+/// ignore: must_be_immutable
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/cupertino/segment_control.dart]: This lint can't be ignored. The fact that widgets are stateless is an invariant that Flutter relies on.

Comment (btobin) [reply examples/flutter_gallery/lib/demo/playground/cupertino/segment_control.dart]: What is a possible solution here? Thanks.

File: examples/flutter_gallery/lib/demo/playground/home.dart [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return DefaultTabController(
+      length: demos.length,
+      child: Scaffold(
+        key: _scaffoldKey,
+        appBar: AppBar(
+          title: Text(title),
+          leading: IconButton(
```
Comment (xster) [inline examples/flutter_gallery/lib/demo/playground/home.dart]: Do you need to do this? Doesn't the app bar do this automatically?

File: examples/flutter_gallery/lib/demo/playground/playground_demo.dart [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+
+  AnimationController _backdropAnimationController;
+
+  @override
+  void initState() {
+    super.initState();
+    _backdropAnimationController = AnimationController(
+        vsync: this,
+        duration: Duration(milliseconds: _animationInterval),
+        value: 1.0);
```
Comment (xster) [inline examples/flutter_gallery/lib/demo/playground/playground_demo.dart]: nit: trailing comma

File: examples/flutter_gallery/lib/demo/playground/playground_demo.dart [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+  }
+
+  Animation<RelativeRect> _getLayerAnimation(BoxConstraints constraints) {
+    final double maxHeight = constraints.biggest.height;
+    final double backPanelHeight = maxHeight - _headerHeight;
+    final double frontPanelHeight = -_headerHeight;
+    const double offsetTop = 130.0;
+
+    return RelativeRectTween(
+            begin: const RelativeRect.fromLTRB(0.0, offsetTop, 0.0, 0.0),
```
Comment (xster) [inline examples/flutter_gallery/lib/demo/playground/playground_demo.dart]: nit: some indent funkiness (trailing comma will fix a bunch of these)

File: examples/flutter_gallery/lib/demo/playground/home.dart (hunk 18 above)
Comment (btobin) [reply examples/flutter_gallery/lib/demo/playground/home.dart]: The `IconButton` does not seem to show automatically in this case. I was surprised as that was the expected behavior. There are existing areas of the gallery codebase that have similar code though. Open to suggestions.

File: examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart (hunk 5 above)
Comment (btobin) [reply examples/flutter_gallery/lib/demo/playground/configuration/material_helpers.dart]: Updated and works, but linter says: 

```[dart] This function has a return type of 'ShapeBorder', but doesn't end with a return statement. [missing_return]```

**Comment by HansMuller** [comment]
I've changed the factoring from functions to widgets, eliminated the "must be immutable" issue, made the demos stateful widgets, fixed much of the formatting, etc.

[playground.zip](https://github.com/flutter/flutter/files/2625871/playground.zip)

File: examples/flutter_gallery/lib/demo/playground/configuration/border_choice.dart [hunk 21]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+class BorderChoice {
+  const BorderChoice({@required this.type, @required this.code});
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/playground/configuration/border_choice.dart]: const BorderChoice({ @required this.type, @required this.code }) : assert(type != null), assert(code != null);

File: examples/flutter_gallery/lib/demo/playground/cupertino/segment_control.dart (hunk 17 above)
Comment (HansMuller) [reply examples/flutter_gallery/lib/demo/playground/cupertino/segment_control.dart]: Sadly, there are loads of conflicts. See https://github.com/flutter/flutter/pull/24764#issuecomment-442526597

**Comment by xster** [comment]
Closing since we did this in https://github.com/flutter/flutter/pull/24842. Let's open a different PR to bring that branch into master.

**Pull request closed without merging**